### PR TITLE
Add FusedGroupedExperts override and offset-aware kernels to skip async EP worst case padding

### PR DESCRIPTION
**Pull request by xmfan**
Add offset-aware SwiGLU kernels to avoid processing overallocated routed tokens buffer

For syncless EP approaches like MinimalAsyncEP and HybridEP

```python
NCCL_DEBUG=WARN  NGPU=4 MODULE=deepseek_v3 CONFIG=deepseek_v3_16b_hybridep wp ./run_train.sh      --parallelism.pipeline_parallel_degree 1       --parallelism.data_parallel_shard_degree 4   --parallelism.expert_parallel_degree 4  --activation_checkpoint.mode full

[rank0]:[titan] 2026-06-13 11:38:47,670 - root - INFO - Initializing HybridEP buffer: hidden_dim=2048, max_tokens_per_rank=16384, num_local_experts=16, ep_size=4
[rank0]:[titan] 2026-06-13 11:39:06,696 - root - INFO - step:  1  loss: 12.04959  grad_norm:  1.7336  memory: 59.90GiB(21.65%)  tps: 441  tflops: 7.98  mfu: 0.32%
[rank0]:[titan] 2026-06-13 11:39:06,697 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:[titan] 2026-06-13 11:39:16,113 - root - INFO - step: 10  loss: 11.20896  grad_norm:  2.5372  memory: 74.71GiB(27.01%)  tps: 15,660  tflops: 283.54  mfu: 11.34%
[rank0]:[titan] 2026-06-13 11:39:24,369 - root - INFO - step: 20  loss:  9.54478  grad_norm:  4.8683  memory: 74.71GiB(27.01%)  tps: 19,847  tflops: 359.36  mfu: 14.37%
[rank0]:[titan] 2026-06-13 11:39:32,798 - root - INFO - step: 30  loss:  8.54311  grad_norm:  2.8289  memory: 74.71GiB(27.01%)  tps: 19,438  tflops: 351.94  mfu: 14.08%
[rank0]:[titan] 2026-06-13 11:39:41,176 - root - INFO - step: 40  loss:  7.89478  grad_norm:  1.2303  memory: 74.71GiB(27.01%)  tps: 19,558  tflops: 354.11  mfu: 14.16%
```

### CODE REVIEW

File: torchtitan/models/common/moe.py [hunk 1]
```diff
@@ -79,18 +89,20 @@ def _experts_forward(
...
+            offs=offsets_E,
         )
-        h_RF = h_RF * torch._grouped_mm(
+        up_RF = torch._grouped_mm(
             x_RD.bfloat16(),
             w3_EFD.bfloat16().transpose(-2, -1),
             offs=offsets_E,
         )
+        if self._use_active_swiglu:
+            h_RF = active_swiglu_op(gate_RF, up_RF, offsets_E[-1:])
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: 1. This seems a `silu_and_mul` op, not swiglu which in addition includes w1 w2 w3 matmul part. So I would suggest renaming.
2. The op seems valuable not only for `GroupedExperts`, but also for `FeedForward` (dense), so a could we make `offset` optional so that it can be shared across.
3. Compared with MinimalAsyncEPTokenDispatcher, the change here is a bit intrusive. I would suggest we combine the kernel into fused w1w3 implementations of feedforward (https://github.com/pytorch/torchtitan/blob/main/torchtitan/overrides/fused_swiglu.py#L69) and grouped experts, and then use `override` to plug in config.

If that's too much, could you just do 1 and 2 in this PR and put the kernel in `torchtitan/overrides` folder? @liangel-02 could take it to 3.

File: torchtitan/overrides/minimal_async_ep_fused_swiglu.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    else:
+        ctx.save_for_backward(gate, up, offsets)
+
+
+silu_and_mul_op.register_autograd(
+    silu_and_mul_autograd_backward, setup_context=silu_and_mul_setup_context
+)
+
+
+class MinimalAsyncEPSiLUAndMulGroupedExperts(GroupedExperts):
```
Comment (xmfan) [inline torchtitan/overrides/minimal_async_ep_fused_swiglu.py]: @tianyu-l I had to add a different override, primarily because the FusedSwiGLU module is a FeedForward.Config override and doesn't have expert dim in the weights

File: torchtitan/overrides/minimal_async_ep_fused_swiglu.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+        GRAD_UP_COL_STRIDE=grad_up.stride(1),
+        BLOCK_M=block_m,
+        BLOCK_N=block_n,
+        num_warps=8,
+    )
+    return grad_gate, grad_up
+
+
+@torch.library.custom_op(
+    "torchtitan_overrides::silu_and_mul",
```
Comment (tianyu-l) [inline torchtitan/overrides/minimal_async_ep_fused_swiglu.py]:
```suggestion
    "torchtitan::silu_and_mul",
```
may be good enough

File: torchtitan/overrides/minimal_async_ep_fused_swiglu.py (hunk 2 above)
Comment (tianyu-l) [inline torchtitan/overrides/minimal_async_ep_fused_swiglu.py]: can we call it
```suggestion
class FusedGroupedExperts(GroupedExperts):
```
so that
- we can fuse w1 and w3 together (can be done in another PR)
- other dispatcher could also enjoy this override

If so, we can also rename the file to e.g. `fused_grouped_experts.py`

Comment (liangel-02) [reply torchtitan/overrides/minimal_async_ep_fused_swiglu.py]: > we can fuse w1 and w3 together (can be done in another PR)

concretely we would call this override [here](https://github.com/pytorch/torchtitan/blob/main/torchtitan/overrides/fused_swiglu.py#L98) right

Comment (tianyu-l) [reply torchtitan/overrides/minimal_async_ep_fused_swiglu.py]: @liangel-02 we should do two things
- let current fused_swiglu.py use this `silu_and_mul` kernel
- enhance `FusedGroupedExperts` with w1, w3 fusion (and add state_dict hooks)

File: torchtitan/overrides/fused_grouped_experts.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+    target=GroupedExperts.Config,
+    description="Use fused SiLU-and-mul activation for MinimalAsyncEP grouped experts.",
+)
+def fused_grouped_experts(
+    cfg: GroupedExperts.Config,
+) -> GroupedExperts.Config:
+    if not isinstance(cfg.token_dispatcher, MinimalAsyncEPTokenDispatcher.Config):
+        return cfg
+
+    warn_once(
```
Comment (xmfan) [inline torchtitan/overrides/fused_grouped_experts.py]: removing this

File: torchtitan/overrides/fused_grouped_experts.py [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+
+silu_and_mul_op.register_autograd(
+    silu_and_mul_autograd_backward, setup_context=silu_and_mul_setup_context
+)
+
+
+class FusedGroupedExperts(GroupedExperts):
+    """Grouped experts variant that uses the override SiLU-and-mul custom op.
+
+    This is intended for ``MinimalAsyncEPTokenDispatcher`` only, where the
```
Comment (xmfan) [inline torchtitan/overrides/fused_grouped_experts.py]: removing this

File: torchtitan/overrides/fused_grouped_experts.py [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+    @dataclass(kw_only=True, slots=True)
+    class Config(GroupedExperts.Config):
+        pass
+
+    def __init__(self, config: Config):
+        super().__init__(config)
+        if not isinstance(self.token_dispatcher, MinimalAsyncEPTokenDispatcher):
+            raise ValueError(
+                "FusedGroupedExperts requires MinimalAsyncEPTokenDispatcher."
+            )
```
Comment (tianyu-l) [inline torchtitan/overrides/fused_grouped_experts.py]: remove?

File: torchtitan/overrides/fused_grouped_experts.py [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+
+silu_and_mul_op.register_autograd(
+    silu_and_mul_autograd_backward, setup_context=silu_and_mul_setup_context
+)
+
+
+class FusedGroupedExperts(GroupedExperts):
+    """Grouped experts variant that uses ``torchtitan::silu_and_mul``.
+
+    This override is currently scoped to ``MinimalAsyncEPTokenDispatcher``,
```
Comment (tianyu-l) [inline torchtitan/overrides/fused_grouped_experts.py]: need to modify docstring accordingly

File: torchtitan/overrides/fused_grouped_experts.py [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+        h_RF = silu_and_mul_op(gate_RF, up_RF, offsets_E)
+        return torch._grouped_mm(
+            h_RF, w2_EDF.bfloat16().transpose(-2, -1), offs=offsets_E
+        ).type_as(x_RD)
+
+
+@override(
+    "fused_grouped_experts",
+    target=GroupedExperts.Config,
+    description="Use fused SiLU-and-mul activation for MinimalAsyncEP grouped experts.",
```
Comment (tianyu-l) [inline torchtitan/overrides/fused_grouped_experts.py]: similar

File: torchtitan/overrides/fused_grouped_experts.py [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+)
+def fused_grouped_experts(
+    cfg: GroupedExperts.Config,
+) -> GroupedExperts.Config:
+    if not isinstance(cfg.token_dispatcher, MinimalAsyncEPTokenDispatcher.Config):
+        return cfg
```
Comment (tianyu-l) [inline torchtitan/overrides/fused_grouped_experts.py]: adapt

File: torchtitan/overrides/fused_grouped_experts.py [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyrefly: ignore-errors
+
+"""
+Fused grouped-experts override.
+
+Importing this module registers the MinimalAsyncEP ``GroupedExperts`` override
+and the ``torchtitan::silu_and_mul`` custom CUDA op it uses.
```
Comment (tianyu-l) [inline torchtitan/overrides/fused_grouped_experts.py]: similar

**Review by tianyu-l — APPROVED** [review]
some more generalization to do

**Review by mlazos — APPROVED** [review]

File: torchtitan/overrides/minimal_async_ep_fused_swiglu.py (hunk 2 above)
Comment (xmfan) [reply torchtitan/overrides/minimal_async_ep_fused_swiglu.py]: btw while enhancing w/ w13 fusion, we will likely just remove `silu_and_mul` and have it in the epilogue, so feel free to take over the code here
